File: code/go/ch09/src/luago/state/api_get.go
package state

import . "luago/api"

// [-0, +1, m]
// http://www.lua.org/manual/5.3/manual.html#lua_newtable
func (self *luaState) NewTable() {
	self.CreateTable(0, 0)
}

// [-0, +1, m]
// http://www.lua.org/manual/5.3/manual.html#lua_createtable
func (self *luaState) CreateTable(nArr, nRec int) {
	t := newLuaTable(nArr, nRec)
	self.stack.push(t)
}

// [-1, +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_gettable
func (self *luaState) GetTable(idx int) LuaType {
	t := self.stack.get(idx)
	k := self.stack.pop()
	return self.getTable(t, k)
}

// [-0, +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_getfield
func (self *luaState) GetField(idx int, k string) LuaType {
	t := self.stack.get(idx)
	return self.getTable(t, k)
}

// [-0, +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_geti
func (self *luaState) GetI(idx int, i int64) LuaType {
	t := self.stack.get(idx)
	return self.getTable(t, i)
}

// push(t[k])
func (self *luaState) getTable(t, k luaValue) LuaType {
	if tbl, ok := t.(*luaTable); ok {
		v := tbl.get(k)
		self.stack.push(v)
		return typeOf(v)
	}

	panic("not a table!") // todo
}

File: code/csharp/ch07/luavm/state/ApiGet.cs
using System;
using LuaType = System.Int32;

namespace luavm.state
{
    partial struct LuaState
    {
        public void NewTable()
        {
            CreateTable(0, 0);
        }

        public void CreateTable(int nArr, int nRec)
        {
            var t = LuaTable.newLuaTable(nArr, nRec);
            stack.push(t);
        }

        public int GetTable(int idx)
        {
            var t = new LuaValue(stack.get(idx)).toLuaTable();
            var k = stack.pop();
            return getTable(new LuaValue(t), new LuaValue(k));
        }

        LuaType getTable(LuaValue t, LuaValue k)
        {
            if (t.isLuaTable())
            {
                var tbl = t.toLuaTable();
                var v = tbl.get(k).value;
                if (v.GetType().IsEquivalentTo(typeof(LuaValue)))
                {
                    v = ((LuaValue) v).value;
                }

                stack.push(v);
                return new LuaValue(v).typeOf();
            }

            throw new Exception("not a table!");
        }

        public LuaType GetField(int idx, string k)
        {
            var t = new LuaValue(stack.get(idx)).toLuaTable();
            return getTable(new LuaValue(t), new LuaValue(k));
//            PushString(k);
//            return GetTable(idx);
        }

        public LuaType GetI(int idx, long i)
        {
            var t = new LuaValue(stack.get(idx)).toLuaTable();
            return getTable(new LuaValue(t), new LuaValue(i));
        }
    }
}